docs/guides/nextjs.md

```diff
@@ -0,0 +1,277 @@
+
+export const CounterStoreContext = createContext()
+
+export interface CounterStoreProviderProps {
+  children: ReactNode
+}
+
+export const CounterStoreProvider = ({
+  children,
+}: CounterStoreProviderProps) => {
+  const storeRef = useRef<CounterStore>()
+  if (!storeRef.current) {
+    storeRef.current = createCounterStore()
+  }
+
+  return (
+    <CounterStoreContext.Provider value={storeRef.current}>
+      {children}
+    </CounterStoreContext.Provider>
+  )
+}
+
+export const useCounterStore = (selector: (store: CounterStore) => T): T => {
+  const counterStoreContext = useContext(CounterStoreContext)
+
+  if (counterStoreContext === undefined) {
+    throw new Error(`useCounterStore must be use within CounterStoreProvider`)
+  }
+
+  return useStore(counterStoreContext, selector)
+}
+```
+
+> **Note:** In this example, we ensure that this component is re-render-safe by checking the
+> value of the reference, so that the store is only created once. This component will only be
+> rendered once per request on the server, but might be re-rendered multiple times on the client if
+> there are stateful client components located above this component in the tree, or if this component
+> also contains other mutable state that causes a re-render.
+
+### Initializing the store
+
+```ts
+// stores/counter-store.ts
+// rest of code
+
+export const initCounterStore = (): CounterState => {
+  return { count: new Date().getFullYear() }
+}
+```
+
+```tsx
+// components/counter-store-provider.tsx
+// rest of code
+
+import {
+  type CounterStore,
+  createCounterStore,
+  initCounterStore,
+} from 'stores/counter-store.ts'
+
+// rest of code
+
+export const CounterStoreProvider = ({
+  children,
+}: CounterStoreProviderProps) => {
+  const storeRef = useRef<CounterStore>()
+  if (!storeRef.current) {
+    storeRef.current = createCounterStore(initCounterStore())
+  }
+
+  return (
+    <CounterStoreContext.Provider value={storeRef.current}>
+      {children}
+    </CounterStoreContext.Provider>
+  )
+}
+
+// rest of code
+```
+
+### Using the store with different architectures
+
+There are two architectures for a Next.js application: the
+[Pages Router](https://nextjs.org/docs/pages/building-your-application/routing) and the
+[App Router](https://nextjs.org/docs/app/building-your-application/routing). The usage of Zustand on
+both architectures should be the same with slight differences related to each architecture.
+
+#### Pages Router
+
+```tsx
+// components/pages/home-page.tsx
+import { useCounterStore } from 'components/counter-store-provider.ts'
+
+export const HomePage = () => {
+  const { count, incrementCount, decrementCount } = useCounterStore()
+
+  return (
+    <div>
+      Count: {count}
+      <hr />
+      <button type="button" onClick={() => void incrementCount()}>
+        Increment Count
+      </button>
+      <button type="button" onClick={() => void decrementCount()}>
+        Decrement Count
+      </button>
```

Comment:
Is it necessary to include the `void` keyword here or can it be omitted because it isn't the subject of the guide?

Reply:
that should be ok for now